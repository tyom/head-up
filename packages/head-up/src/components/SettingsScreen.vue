<template>
  <ModalDialogue
    :show="show"
    :heading="heading"
    @close="$emit('close')"
  >
    <form class="settings">
      <p>
        <VSwitchToggle v-model="smoothScrollingValue" label="Smooth scrolling"/>
      </p>
    </form>
  </ModalDialogue>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import ModalDialogue from './ModalDialogue';

export default {
  components: {
    ModalDialogue,
  },
  props: {
    show: {
      type: Boolean,
      default: false,
    },
    heading: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      smoothScrollingValue: this.smoothScrolling,
    };
  },
  computed: {
    ...mapState(['smoothScrolling']),
  },
  watch: {
    // TODO This appears to be some async issue as the value should be
    // initialised in data function
    smoothScrolling: {
      immediate: true,
      handler(newVal) {
        this.smoothScrollingValue = newVal;
      },
    },
    smoothScrollingValue(newVal) {
      this.UPDATE_SETTING({
        smoothScrolling: newVal,
      });
    },
  },
  methods: {
    ...mapActions(['UPDATE_SETTING']),
  },
};
</script>
