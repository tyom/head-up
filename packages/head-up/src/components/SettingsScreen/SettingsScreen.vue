<template>
  <form class="settings">
    <ul>
      <li v-for="(setting, key) in settings" :key="key">
        <VSwitchToggle
          v-model="settingsState[key].value"
          :label="setting.label"
          :name="key"
        />
      </li>
    </ul>
  </form>
</template>

<script>
export default {
  props: {
    settings: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      settingsState: this.settings,
    };
  },
  watch: {
    settingsState: {
      deep: true,
      handler() {
        this.$emit('save', this.settingsState);
      },
    },
  },
};
</script>

<style scoped>
ul {
  list-style: none;
  padding: 0;
  margin: 0;
}

li + li {
  margin-top: 1em;
}
</style>
