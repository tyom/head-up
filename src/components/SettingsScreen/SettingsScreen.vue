<template>
  <div class="settings">
    <ul>
      <li v-for="(setting, key) in settings" :key="key">
        <VSwitchToggle
          v-model="settings[key].value"
          :label="setting.label"
          :name="key"
        />
      </li>
      <li>
        <button type="button" class="reset-button" @click="handleResetState">
          Reset state
        </button>
      </li>
    </ul>
  </div>
</template>

<script>
import store from '../../store';

export default {
  data() {
    return {
      settings: { ...store.state.settings },
    };
  },
  watch: {
    settings: {
      deep: true,
      handler(newSettings) {
        store.dispatch('UPDATE_SETTINGS', newSettings);
      },
    },
  },
  methods: {
    handleResetState() {
      store.dispatch('RESET_STATE');
    },
  },
};
</script>

<style scoped>
ul {
  list-style: none;
  padding: 0;
  margin: 0;
}

li + li {
  margin-top: 1em;
}
</style>
