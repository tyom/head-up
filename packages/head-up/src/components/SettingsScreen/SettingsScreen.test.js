import { shallowMount } from '@vue/test-utils';
import SettingsScreen from './SettingsScreen';

test('render default', () => {
  const wrapper = shallowMount(SettingsScreen, {
    propsData: {
      settings: {
        myFavouriteSetting: {
          value: true,
          label: 'Setting I prefer',
        },
      },
    },
    stubs: ['VSwitchToggle'],
  });
  expect(wrapper).toMatchSnapshot();
});

test('emits save event', () => {
  const wrapper = shallowMount(SettingsScreen, {
    propsData: {
      settings: {
        myFavouriteSetting: {
          value: true,
          label: 'Setting I prefer',
        },
      },
    },
    stubs: ['VSwitchToggle'],
  });

  wrapper.vm.settingsState.myFavouriteSetting.value = false;
  expect(wrapper.emitted('save')[0][0]).toHaveProperty(
    'myFavouriteSetting.value',
    false,
  );
});
