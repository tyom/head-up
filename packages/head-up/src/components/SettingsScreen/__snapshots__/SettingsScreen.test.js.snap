// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<form class="settings">
  <ul>
    <li>
      <vswitchtoggle-stub label="Setting I prefer"></vswitchtoggle-stub>
    </li>
  </ul>
</form>
`;
